File: day06.py
#!/usr/bin/env python3

from get_input import get_input
import re
from itertools import product

instructions = re.compile(
        r'(toggle|turn off|turn on) (\d+),(\d+) through (\d+),(\d+)')

def apply(grid, func, range_a, range_b):
    for m, n in product(range_a, range_b):
        grid[m][n] = func(grid[m][n])
    return grid

def test_apply():
    grid = [[0 for _ in range(1000)] for _ in range(1000)]
    grid = apply(grid, 'turn on', range(0,1000), range(0,1))
    assert sum(sum(row) for row in grid) == 1000

def part1(lines):
    def modes(mode):
        if mode == 'turn on':
            return lambda _: 1
        elif mode == 'turn off':
            return lambda _: 0
        elif mode == 'toggle':
            return lambda v: 0 if v == 1 else 1
        else:
            raise ValueError("Not a valid instruction {}".format(mode))

    grid = [[0 for _ in range(1000)] for _ in range(1000)]

    for line in lines:
        m = instructions.match(line)
        if m is None:
            raise ValueError("Not a valid line {}".format(line))
        mode, args = m.group(1), [int(arg) for arg in m.groups()[1:]]
        grid = apply(grid, modes(mode), range(args[0], args[2]+1), 
                range(args[1], args[3]+1))
    return sum(sum(row) for row in grid)

def part2(lines):
    def modes(mode):
        if mode == 'turn on':
            return lambda v: v + 1
        elif mode == 'turn off':
            return lambda v: v - 1 if v > 0 else 0
        elif mode == 'toggle':
            return lambda v: v + 2
        else:
            raise ValueError("Not a valid instruction {}".format(mode))

    grid = [[0 for _ in range(1000)] for _ in range(1000)]

    for line in lines:
        m = instructions.match(line)
        if m is None:
            raise ValueError("Not a valid line {}".format(line))
        mode, args = m.group(1), [int(arg) for arg in m.groups()[1:]]
        grid = apply(grid, modes(mode), range(args[0], args[2]+1), 
                range(args[1], args[3]+1))
    return sum(sum(row) for row in grid)

if __name__ == '__main__':
    lines = get_input(day=6, year=2015).splitlines()
    print("Part 1: {}".format(part1(lines)))
    print("Part 2: {}".format(part2(lines)))

File: day06.rs
use common::get_input;
use std::collections::HashMap;
use regex::Regex;

type Point = (usize, usize);

#[derive(Debug)]
enum Action {
    TurnOff,
    TurnOn,
    Toggle,
}

#[derive(Debug)]
struct Instruction {
    action: Action,
    from: Point,
    to: Point,
}

impl Instruction {
    fn from_string(text: String) -> Instruction {
        let re = Regex::new(r"^(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)").unwrap();
        let captures = re.captures(&text).unwrap();

        let get_int = |n: usize| -> usize {
            captures.get(n).map(|c| c.as_str().parse::<usize>().unwrap()).unwrap()
        };

        let action = match captures.get(1).unwrap().as_str() {
            "toggle" => Action::Toggle,
            "turn on" => Action::TurnOn,
            "turn off" => Action::TurnOff,
            _ => unimplemented!(),
        };

        Instruction {
            action,
            from: (get_int(2), get_int(3)),
            to: (get_int(4), get_int(5)),
        }
    }
}

type Input = Vec<Instruction>;

fn part1(instructions: &Input) -> usize {
    let mut grid: HashMap<Point, bool> = HashMap::new();
    for instruction in instructions {
        for row in instruction.from.0..(instruction.to.0+1) {
            for col in instruction.from.1..(instruction.to.1+1) {
                let point = (row, col);
                match instruction.action {
                    Action::TurnOn => { grid.insert(point, true); }
                    Action::TurnOff => { grid.insert(point, false); },
                    Action::Toggle => { grid.entry((row, col)).and_modify(|v| *v = !*v).or_insert(!false); },
                };
            }
        }
    };
    grid.values().filter(|&&v| v).count()
}

fn part2(instructions: &Input) -> usize {
    let mut grid: HashMap<Point, usize> = HashMap::new();
    for instruction in instructions {
        for row in instruction.from.0..(instruction.to.0+1) {
            for col in instruction.from.1..(instruction.to.1+1) {
                let point = (row, col);
                match instruction.action {
                    Action::TurnOn => *grid.entry(point).or_insert(0) += 1,
                    Action::TurnOff => if let Some(v) = grid.get_mut(&point) { if *v > 0 { *v -= 1 }},
                    Action::Toggle => *grid.entry(point).or_insert(0) += 2,
                };
            }
        }
    };
    grid.values().sum()
}

fn parse(text: String) -> Input {
    text.trim().split("\n").map(|i| Instruction::from_string(i.to_string())).collect()
}

fn main() {
    let input = parse(get_input(06, 2015));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
